client/src/main/scala/io/delta/sharing/client/auth/OAuthClientCredentialsAuthProvider.scala

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.sharing.client.auth
+
+import org.apache.http.HttpHeaders
+import org.apache.http.client.methods.HttpRequestBase
+import org.apache.http.impl.client.CloseableHttpClient
+
+import io.delta.sharing.client.OAuthClientCredentialsDeltaSharingProfile
+
+private[client] case class OAuthClientCredentialsAuthProvider(
+                                           client: CloseableHttpClient,
+                                           profile: OAuthClientCredentialsDeltaSharingProfile)
+  extends AuthCredentialProvider {
+
+  private[auth] lazy val oauthClient = new OAuthClient(client,
+    profile.tokenEndpoint, profile.clientId, profile.clientSecret, profile.scope)
+
+  private var currentToken: Option[OAuthClientCredentials] = None
+
+  override def addAuthHeader(httpRequest: HttpRequestBase): Unit = {
+    if (currentToken.isEmpty || needsRefresh(currentToken.get)) {
+      maybeRefreshToken()
+    }
```

Comment:
Perhaps directly call maybeRefreshToken, since these conditions are anyways checked there and this access is not synchronized so there may be concurrent access issues.

Reply:
The synchronization logic in the PR with [Double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking) was correct. However for better readability I changed that to use a reentrant lock. That should improve readability.